From 2917e4144fd27ceb840235bf587c40d66fb6dff8 Mon Sep 17 00:00:00 2001
From: l4h <z@univalence.me>
Date: Sat, 17 Oct 2026 23:08:26 -0700
Subject: [PATCH] x86_64: add seL4_X86_ExecuteNever for user mappings

seL4 on x86 maps every readable user page executable: there is no way to
map data without execute permission, so a system that runs untrusted code
cannot enforce W^X for it.

Add a VM attribute, seL4_X86_ExecuteNever (bit 3), that sets the XD bit
of 4 KiB, 2 MiB and 1 GiB user mappings. Boot code enables EFER.NXE on
every core when the CPU has execute-disable; without it the attribute is
ignored, as XD is a reserved bit then.
---
 include/arch/x86/arch/64/mode/kernel/vspace.h   |  3 +++
 .../arch/x86/arch/64/mode/object/structures.bf  |  3 ++-
 libsel4/arch_include/x86/sel4/arch/types.h      |  3 +++
 src/arch/x86/64/head.S                          | 12 +++++++++++-
 src/arch/x86/64/kernel/vspace.c                 | 17 ++++++++++++++---
 src/arch/x86/kernel/vspace.c                    |  6 ++++++
 6 files changed, 39 insertions(+), 5 deletions(-)

diff --git a/include/arch/x86/arch/64/mode/kernel/vspace.h b/include/arch/x86/arch/64/mode/kernel/vspace.h
index 413843e75..127b49a7a 100644
--- a/include/arch/x86/arch/64/mode/kernel/vspace.h
+++ b/include/arch/x86/arch/64/mode/kernel/vspace.h
@@ -14,6 +14,9 @@ struct lookupPDPTSlot_ret {
 };
 typedef struct lookupPDPTSlot_ret lookupPDPTSlot_ret_t;
 
+/* Whether execute-disable is enabled (seL4_X86_ExecuteNever mappings). */
+bool_t x86_64_nxEnabled(void);
+
 static inline pte_t x86_make_device_pte(paddr_t phys)
 {
     return pte_new(
diff --git a/include/arch/x86/arch/64/mode/object/structures.bf b/include/arch/x86/arch/64/mode/object/structures.bf
index d24a7d032..2320af061 100644
--- a/include/arch/x86/arch/64/mode/object/structures.bf
+++ b/include/arch/x86/arch/64/mode/object/structures.bf
@@ -265,7 +265,8 @@ block VMFault {
 -- VM attributes
 
 block vm_attributes {
-    padding         61
+    padding         60
+    field x86XDBit  1
     field x86PATBit 1
     field x86PCDBit 1
     field x86PWTBit 1
diff --git a/libsel4/arch_include/x86/sel4/arch/types.h b/libsel4/arch_include/x86/sel4/arch/types.h
index c8e2de6c5..8fb8635eb 100644
--- a/libsel4/arch_include/x86/sel4/arch/types.h
+++ b/libsel4/arch_include/x86/sel4/arch/types.h
@@ -34,6 +34,9 @@ typedef enum {
     seL4_X86_CacheDisabled = 2,
     seL4_X86_Uncacheable = 3,
     seL4_X86_WriteCombining = 4,
+    /* x86_64: the mapping is not executable (XD), where the CPU supports
+     * it. May be combined with the cache attributes. */
+    seL4_X86_ExecuteNever = 8,
     SEL4_FORCE_LONG_ENUM(seL4_X86_VMAttributes),
 } seL4_X86_VMAttributes;
 
diff --git a/src/arch/x86/64/head.S b/src/arch/x86/64/head.S
index f489757fb..cf3e85ed0 100644
--- a/src/arch/x86/64/head.S
+++ b/src/arch/x86/64/head.S
@@ -231,11 +231,21 @@ BEGIN_FUNC(enable_x64_mode)
     movl %cr4, %eax
     orl $0x20, %eax
     movl %eax, %cr4
-    /* Set LME (bit 8) in the extended feature MSR. */
+    /* Set LME (bit 8) in the extended feature MSR, and NXE (bit 11) if
+     * the CPU has execute-disable (CPUID 0x80000001 EDX bit 20), for
+     * seL4_X86_ExecuteNever mappings. */
+    pushl %ebx
+    movl $0x80000001, %eax
+    cpuid
+    andl $0x100000, %edx
+    shrl $9, %edx
+    movl %edx, %ebx
     movl $IA32_EFER_MSR, %ecx
     rdmsr
     orl $0x100, %eax
+    orl %ebx, %eax
     wrmsr
+    popl %ebx
     /* Set PG (bit 31) of cr0 to enable paging. */
     movl %cr0, %eax
     orl $0x80000000, %eax
diff --git a/src/arch/x86/64/kernel/vspace.c b/src/arch/x86/64/kernel/vspace.c
index bab7edd57..9bdde4b3b 100644
--- a/src/arch/x86/64/kernel/vspace.c
+++ b/src/arch/x86/64/kernel/vspace.c
@@ -841,6 +841,17 @@ static pml4e_t CONST makeUserPML4E(paddr_t paddr, vm_attributes_t vm_attr)
            );
 }
 
+/* Whether boot code enabled execute-disable (EFER.NXE): the CPU has it. */
+bool_t x86_64_nxEnabled(void)
+{
+    static int nxe = -1;
+
+    if (nxe < 0) {
+        nxe = !!(x86_rdmsr(IA32_EFER_MSR) & BIT(11));
+    }
+    return nxe;
+}
+
 static pml4e_t CONST makeUserPML4EInvalid(void)
 {
     return pml4e_new(
@@ -858,7 +869,7 @@ static pml4e_t CONST makeUserPML4EInvalid(void)
 static pdpte_t CONST makeUserPDPTEHugePage(paddr_t paddr, vm_attributes_t vm_attr, vm_rights_t vm_rights)
 {
     return pdpte_pdpte_1g_new(
-               0,          /* xd               */
+               vm_attributes_get_x86XDBit(vm_attr), /* xd       */
                paddr,      /* physical address */
                0,          /* PAT              */
                0,          /* global           */
@@ -903,7 +914,7 @@ static pdpte_t CONST makeUserPDPTEInvalid(void)
 pde_t CONST makeUserPDELargePage(paddr_t paddr, vm_attributes_t vm_attr, vm_rights_t vm_rights)
 {
     return pde_pde_large_new(
-               0,                                              /* xd                   */
+               vm_attributes_get_x86XDBit(vm_attr),           /* xd                   */
                paddr,                                          /* page_base_address    */
                vm_attributes_get_x86PATBit(vm_attr),          /* pat                  */
                0,                                              /* global               */
@@ -952,7 +963,7 @@ pde_t CONST makeUserPDEInvalid(void)
 pte_t CONST makeUserPTE(paddr_t paddr, vm_attributes_t vm_attr, vm_rights_t vm_rights)
 {
     return pte_new(
-               0,                                              /* xd                   */
+               vm_attributes_get_x86XDBit(vm_attr),           /* xd                   */
                paddr,                                          /* page_base_address    */
                0,                                              /* global               */
                vm_attributes_get_x86PATBit(vm_attr),          /* pat                  */
diff --git a/src/arch/x86/kernel/vspace.c b/src/arch/x86/kernel/vspace.c
index ddf4ce451..8ca477b79 100644
--- a/src/arch/x86/kernel/vspace.c
+++ b/src/arch/x86/kernel/vspace.c
@@ -962,6 +962,12 @@ exception_t decodeX86FrameInvocation(
         vaddr = getSyscallArg(0, buffer);
         w_rightsMask = getSyscallArg(1, buffer);
         vmAttr = vmAttributesFromWord(getSyscallArg(2, buffer));
+#ifdef CONFIG_ARCH_X86_64
+        /* Without execute-disable the XD bit is reserved: ignore it. */
+        if (vm_attributes_get_x86XDBit(vmAttr) && !x86_64_nxEnabled()) {
+            vmAttr = vm_attributes_set_x86XDBit(vmAttr, 0);
+        }
+#endif
         vspaceCap = current_extra_caps.excaprefs[0]->cap;
 
         capVMRights = cap_frame_cap_get_capFVMRights(cap);
-- 
2.47.3

